pilot/pkg/model/mesh.go

```diff
@@ -0,0 +1,785 @@
+	// assumed to be trusted within the entire mesh, irrespective of which
+	// platform, cluster or availability zone this instance runs in.
+	USER destinationAttribute = "destination.user"
+)
+
+var (
+	destinationAttrSet = map[string]bool{
+		UID.stringValue():       true,
+		SERVICE.stringValue():   true,
+		NAME.stringValue():      true,
+		NAMESPACE.stringValue(): true,
+		IP.stringValue():        true,
+		PORT.stringValue():      true,
+		DOMAIN.stringValue():    true,
+		USER.stringValue():      true,
+	}
+
+	canonicalIstioDomain = "svc.cluster.local"
+	svcRegexp            = regexp.MustCompile(
+		"([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.svc\\.cluster\\.local")
+	domainRegexp = regexp.MustCompile(
+		"([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])*")
+)
+
+type destinationAttribute string
+
+// Mesh is a platform independent abstraction used by Controllers
+// for maintaining a list of service endpoints used by this Pilot.
+//
+// Controllers under https://github.com/istio/istio/tree/master/pilot/platform
+// update the list of service endpoints in this Mesh with those
+// available in a platform specific service registry.
+//
+// Under the hoods, the Mesh implements the necessary logic required for endpoint
+// discovery based on routing rules and endpoint subsets.
+// See https://github.com/istio/api/search?q=in%3Afile+"message+Subset"+language%3Aproto
+// This logic includes comparing the updated list provided by Controllers with
+// what this view holds and accordingly updating internal structures used for
+// endpoint discovery.
+type Mesh struct {
+
+	// allEndpoints is a map of the UID to the server instance. The UID is expected
+	// to be unique across the mesh. Mesh uses this map during updates from
+	// Controllers to locate the previous representation  of the endpoint and check
+	// for changes to its metadata that could result in changes to the subsetEndpoints.
+	allEndpoints map[string]*Endpoint
+
+	// subsetDefinitions holds the metadata, basically the set of lable name values
+	// that ought to match corresponding attributes of the endpoint.
+	// The actual subset of allEndpoints that match these subset definitions is
+	// stored in subsetEndpoints. Mesh uses this map for identifying config changes
+	// to the subset's metadata. The key of this map is the Subset name.
+	subsetDefinitions map[string]*route.Subset
+
+	// subsetEndpoints allows for look-ups from the Subset name to a set of Endpoint UIDs.
+	// The UIDs in the set are guaranteed to exist in allEndpoints and the
+	// corresponding LbEndpointis guaranteed to satisfy the labels in the corresponding
+	// Subset.
+	subsetEndpoints subsetEndpoints
+
+	// reverseAttrMap provides reverse lookup from label name > label value >
+	// endpointUIDSet. All Endpoints in the set are guaranteed to match the
```

Comment:
what label is this?
and isn't name->value what a label is supposed to be? So I don't understand the "reverse" aspect of this.

Reply:
labels from a subset ...  It's a quick way to figure out impacts to config changes based on a large number of endpoints. In any case, this is an implementation detail which we can address in subsequent PRs. If anyone would like me to spend the time I can compare performance numbers for the current aggregation and one using reverse maps. If it turns out that reverse maps works better than O(N^3) or possibly even O(N^4), I'll be happy to nuke reverse maps from this code and revert back to 4 levels of 4 loops

Reply:
This is still wrong btw. Please read the route rule api. Subsets are specific per service. We do not compute subsets solely based on labels. It’s a combination of service name and label set.

Reply:
And once again, since you invoked the speed and efficiency argument, this two three level hash traversal is super cache unfriendly. Concatenate servicename.labelName.labelValue into a single string. Use this string as the key for the map. The value is an array of endpoint pointers (not yet another lookup using uid). From the endpoint pointers, you get the endpoint struct and get into the owning clusters directly.

Reply:
@rshriram everything here does not have to exactly match route rules. Everything in the PR doesn't even have to match public interfaces in this PR exactly. This is internal implementation that can change. And it should change based on metrics and hard evidence. As I've said before if I find evidence that this internal implementation is less performant, perf tests will course correct that.